src/iptvsimple/ConnectionManager.cpp

```diff
@@ -0,0 +1,181 @@
+/*
+ *  Copyright (C) 2005-2021 Team Kodi (https://kodi.tv)
+ *
+ *  SPDX-License-Identifier: GPL-2.0-or-later
+ *  See LICENSE.md for more information.
+ */
+
+#include "ConnectionManager.h"
+
+#include "IConnectionListener.h"
+#include "InstanceSettings.h"
+#include "utilities/Logger.h"
+#include "utilities/WebUtils.h"
+
+#include <chrono>
+
+#include <kodi/tools/StringUtils.h>
+#include <kodi/Network.h>
+
+using namespace iptvsimple;
+using namespace iptvsimple::utilities;
+using namespace kodi::tools;
+
+/*
+ * Iptvsimple Connection handler
+ */
+
+ConnectionManager::ConnectionManager(IConnectionListener& connectionListener, std::shared_ptr<iptvsimple::InstanceSettings> settings)
+  : m_connectionListener(connectionListener), m_settings(settings), m_suspended(false), m_state(PVR_CONNECTION_STATE_UNKNOWN)
+{
+}
+
+ConnectionManager::~ConnectionManager()
+{
+  Stop();
+}
+
+void ConnectionManager::Start()
+{
+  // Note: "connecting" must only be set one time, before the very first connection attempt, not on every reconnect.
+  SetState(PVR_CONNECTION_STATE_CONNECTING);
+  m_running = true;
+  m_thread = std::thread([&] { Process(); });
+}
+
+void ConnectionManager::Stop()
+{
+  m_running = false;
+  if (m_thread.joinable())
+    m_thread.join();
+
+  Disconnect();
+}
+
+void ConnectionManager::OnSleep()
+{
+  std::lock_guard<std::mutex> lock(m_mutex);
+
+  Logger::Log(LogLevel::LEVEL_DEBUG, "%s going to sleep", __func__);
+
+  m_suspended = true;
+}
+
+void ConnectionManager::OnWake()
+{
+  std::lock_guard<std::mutex> lock(m_mutex);
+
+  Logger::Log(LogLevel::LEVEL_DEBUG, "%s Waking up", __func__);
+
+  m_suspended = false;
+}
+
+void ConnectionManager::SetState(PVR_CONNECTION_STATE state)
+{
+  PVR_CONNECTION_STATE prevState(PVR_CONNECTION_STATE_UNKNOWN);
+  PVR_CONNECTION_STATE newState(PVR_CONNECTION_STATE_UNKNOWN);
+
+  {
+    std::lock_guard<std::mutex> lock(m_mutex);
+
+    /* No notification if no state change or while suspended. */
+    if (m_state != state && !m_suspended)
+    {
+      prevState = m_state;
+      newState = state;
+      m_state = newState;
+
+      Logger::Log(LogLevel::LEVEL_DEBUG, "connection state change (%d -> %d)", prevState, newState);
+    }
+  }
+
+  if (prevState != newState)
+  {
+    static std::string serverString;
+
+    if (newState == PVR_CONNECTION_STATE_SERVER_UNREACHABLE)
+    {
+      m_connectionListener.ConnectionLost();
+    }
+    else if (newState == PVR_CONNECTION_STATE_CONNECTED)
+    {
+      m_connectionListener.ConnectionEstablished();
+    }
+
+    /* Notify connection state change (callback!) */
+    m_connectionListener.ConnectionStateChange(m_settings->GetM3ULocation(), newState, "");
```

Comment:
the notification may be slightly annoying? Do you think its required in this instance?

Reply:
Ok, moved this behind a variable for now. We may want this in the future if we extend functionality.

Reply:
Reverted this change as the state change is required.